### PR TITLE
[Feature] Add Task.requestJSONEncodable

### LINKED ISSUE #1325: Add Task case to handle request with Encodable type.
Long read: #1135. Short one: we need an option to send `Encodable` type that will be converted to data and passed to `httpBody`. It could look somehow similar to:
```swift
enum Task {
    ...
    case requestJsonEncodable(Encodable)
    ...
}
```

### PR DESCRIPTION
**Pull request by sunshinejr**
Resolves #1325. We had 2 PRs for this one, but both needed some enhancements and because we want to release `10.0.0-beta.1` in the near future, I decided to finish it myself while giving credits to both @LeLuckyVint and @afonsograca. Thank you guys **_so_** much because without you we wouldn't have this feature that fast in the next release!

Additionally, @afonsograca is already in @Moya/contributors, but @LeLuckyVint is not, so I invited you manually as well. Would love you guys to review this one!

### CODE REVIEW

File: docs/Examples/ErrorTypes.md [hunk 1]
```diff
@@ -33,9 +33,11 @@ case .statusCode(let response):
     print(response)
 case .stringMapping(let response):
     print(response)
-case objectMapping(let error, let response):
+case .objectMapping(let error, let response):
```
Comment (devxoul) [inline docs/Examples/ErrorTypes.md]: Thanks!

File: docs/Endpoints.md [hunk 2]
```diff
@@ -109,12 +109,11 @@ That's what the `requestClosure` parameter is for.
...
-method throws two possible errors: `MoyaError.requestMapping(String)` error in the 
-case that an `URLRequest` could not be created for the given path and a `Swift.Error`
-in the case of an error related to the encoding of the parameters for a request. 
-This `Swift.Error` is wrapped up as a `MoyaError.parameterEncoding(Swift.Error)` by the 
-`MoyaProvider.defaultRequestMapping` closure. 
+which uses the `urlRequest()` method of the `Endpoint` instance. This `urlRequest()` 
+method throws three possible errors: 
+- `MoyaError.requestMapping(String)` when `URLRequest` could not be created for given path
+- `MoyaError.parameterEncoding(Swift.Error)` when parameters couldn't be encoded
+- `MoyaError.encodableMapping(Swift.Error)` when `Encodable` object couldn't be encoded into `Data`
```
Comment (SD10) [inline docs/Endpoints.md]: Much cleaner 👍

File: docs_CN/Targets.md [hunk 3]
```diff
@@ -77,6 +77,7 @@ public var task: Task {
  
 - `.requestPlain` 没有任何东西发送
 - `.requestData(_:)` 可以发送 `Data` (useful for `Encodable` types in Swift 4)
+- `.requestJSONEncodable(_:)`
 - `.requestParameters(parameters:encoding:)` 发送指定编码的参数
```
Comment (SD10) [inline docs_CN/Targets.md]: Bonus points for updating the Chinese docs 😱

File: Changelog.md [hunk 4]
```diff
@@ -2,6 +2,7 @@
 ### Added
 - **Breaking Change** `.parameterEncoding(Swift.Error)` case to `MoyaError`. [#1248](https://github.com/Moya/Moya/pull/1248) by [@SD10](https://github.com/SD10).
 - **Breaking Change** `Decodable` object mapping methods to `Moya.Response`. [#1335](https://github.com/Moya/Moya/pull/1335) by [@devxoul](https://github.com/devxoul).
+- `.requestJSONEncodable` case to  `Task`. [#1349](https://github.com/Moya/Moya/pull/1349) by [LeLuckyVint](https://github.com/LeLuckyVint), [@afonsograca](https://github.com/afonsograca) and [@sunshinejr](https://github.com/sunshinejr).
```
Comment (pedrovereza) [inline Changelog.md]: You missed the `@` when tagging @LeLuckyVint :wink:

**Comment by SD10** [comment]
I think this is great and a good way to handle both PRs. I do think this is better being listed as a breaking change because you are breaking the exhaustiveness of two enums 🤔

**Comment by sunshinejr** [comment]
True @SD10, we should probably list addition of new error cases as well.

File: Changelog.md (hunk 4 above)
Comment (sunshinejr) [reply Changelog.md]: :D you always spot my little errors that nobody notices, thanks!

**Comment by SD10** [comment]
@sunshinejr Yeah, I didn't know if the policy was to keep a single entry per PR or to add multiple entries when dealing with an addition + a breaking change. I notice we didn't mention the new error case in #1335 as well.

**Comment by sunshinejr** [comment]
@pedrovereza @SD10 pushed my fixes, let me know how it looks now!

**Comment by sunshinejr** [comment]
Just a heads-up that I'll probably merge this later today as we plan to release 10.0.0-beta.1. So if anyone can/want to review this one, it will be greatly appreciated!

**Pull request merged**